lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -214,6 +214,7 @@ void ThreadPoolExecutor::handleDeviceManagerResult(
     // will transfer. //executionState->transferOutputs();
     ResultCBTy cb = executionState->getCallback();
     DCHECK(cb != nullptr);
+    executionState->returnToPool();
```

Comment:
If there a way to use some sort of RAII to make sure we return it no matter what?

Reply:
One pattern that I've seen working well is to just have the pool give out the NetworkExecutionState wrapped in a unique_ptr. Then have the user return the unique_ptr back to the pool once it's done. Basically this models "whoever has the pointer has ownership". It doesn't do RAII to ensure we always return to the pool though.

One potential blocker to do this is that in c++ 14, std::function can't wrap a lambda that captures a unique_ptr because unique_ptr isn't copyable. folly::Function is a better solution that handles this case correctly. https://github.com/facebook/folly/blob/master/folly/docs/Function.md

Reply:
I don't think we can use a unique_ptr here because execution can fork. If there are two sibling nodes in the DAG both calls of executeDAGNode need a reference to the state. So we could use a shared_ptr but that doesn't enforce ownership.

Reply:
I see. Good point. raw pointer is good then. It's better than shared pointer.